packages/react-reconciler/src/ReactUpdateQueue.js

```diff
@@ -410,6 +410,16 @@ function getStateFromUpdate<State>(
           }
         }
         partialState = payload.call(instance, prevState, nextProps);
+        if (__DEV__ && workInProgress.mode & StrictMode) {
+          if (partialState === undefined) {
+            warning(
+              false,
+              'A setState updater function was called that returned undefined. ' +
+                'Check if you forgot to return a value in your setState updater ' +
```

Comment:
This message is a good start (though I also agree with @nhunzaker above). However, it misses the opportunity to provide more directed/clearer guidance for fixing the common mistake you cited in #13121:

> this.setState(({ bool }) => { bool: !bool });

i.e. they intended to use an arrow function with a concise body (implicit return), but they forgot to wrap the object literal in parenthesis to turn it into a JS expression.

If I was presented with the above, it might still take me a minute to realize what the actual issue is. 🤔

I think something along these lines would help with that:
> This could happen if you are using an arrow function to return an object literal expression, but forgot to wrap the object literal in parenthesis.

Reply:
I like that but I should also point out that this is also very likely to happen if you do 

```js
this.setState(state => {
    // some logic here
    // forgot to return
})
```

how to address that as well?

Reply:
I didn't mean to get rid of ln 419, meaning you could still say, "... Otherwise, if a no-op was intended, return null."

In sum, maybe:
> A setState updater function was called that returned undefined. This could happen if you are using an arrow function to return an object literal expression, but forgot to wrap the object literal in parenthesis. Otherwise, if a no-op was intended, ensure that you return null within the updater function body.

Reply:
I think in the past, we've linked out to reference material to break down a problem more clearly. Is there a page that walks through setState's updater function usage that we could reference?

@gaearon do you know if this is still accurate, and applicable in this case?

Reply:
here? https://reactjs.org/docs/react-component.html#setstate

Reply:
@sw-yx Just reread [your comment](https://github.com/facebook/react/pull/13121#discussion_r199158389) and I see what you mean; the warning doesn't address the case where the user *intended* to return non-null, but simply forgot.

I guess we'd have to say something like (which is now getting a little verbose):
> A setState updater function was called that returned undefined. This could happen if you are using an arrow function to return an object literal expression, but forgot to wrap the object literal in parenthesis. **Otherwise, ensure that you either return an object or null (if no update is required).**